Enable coupled GOCART

rmontuoro opened the pull request:
This PR introduces preliminary changes to enable running the [FV3-GOCART](https://github.com/NOAA-EMC/FV3-GOCART) (UFS-Aerosols) app via the coupled global workflow on Hera and Orion. The app integrates a new UFS aerosols component (under development) directly coupled to the atmospheric model. 

Two test cases are also included as templates to help set up free coupled aerosols forecast runs.

This work supports the integration of NASA's 2nd-generation GOCART model component into the fully-coupled UFS Weather Model. See issue #304.

rmontuoro committed:
Add temporary changes to run the UFS-GOCART application.

rmontuoro committed:
Update README.md

rmontuoro committed:
Add GRDFIX and OROFIX places settings to allow using
custom locations for gridspec and orography files,
respectively.

rmontuoro committed:
Add CFGDIR and INPDIR chemistry settings to specify
the directory path for the required model resource
and data input files, respectively.

rmontuoro committed:
Minor cleanup.

rmontuoro committed:
Add case template for free coupled aerosol forecast
over the FIREX period.

rmontuoro committed:
Merge branch 'feature/aerosols' of https://github.com/rmontuoro/global-workflow into feature/aerosols

rmontuoro committed:
Enable CFGDIR and INPDIR.

rmontuoro committed:
Add number of diagnostic tracers (ntdiag) to chemistry settings.

rmontuoro committed:
Update FV3 field and diagnostic tables to reorder aerosols.

Update test case files for coupled aerosol runs.

rmontuoro committed:
Update README.md

rmontuoro committed:
Properly retrieve optional string values.

rmontuoro committed:
Load proper modules for coupled aerosol runs on Orion.

rmontuoro committed:
Update coupled aerosols template cases.

JessicaMeixner-NOAA commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# How to use the unified workflow for the coupled ufs-weather-model (work in progress)
+# How to use the unified workflow for the coupled ufs-aerosol app (work in progress)
```
??? Feature/coupled-crow is now being changed to the aerosol app?

JessicaMeixner-NOAA replied:
Change to: 

How to use the unified (or feature/coupled-crow) workflow for the coupled ufs-weather-model or the aersol app

JessicaMeixner-NOAA commented on README.md:
```diff
@@ -7,13 +7,13 @@ cd coupled-workflow
 git submodule update --init --recursive     #Update submodules 
 cd sorc
-sh checkout.sh -c                           #Check out forecast model with COUPLED=YES
+sh checkout.sh -a                           #Check out forecast model with active aerosols
```
Unless all the coupled model development is leaving the ufs-weather-model and moving to be used to the aerosol model, can the README please be modified so that this shows options for "coupled" or "aerosol" instead of aerosol only?

JessicaMeixner-NOAA commented on README.md:
````diff
@@ -48,25 +48,19 @@ Then, open and edit user.yaml:
+./setup_case.sh -p HERA ../cases/aerosol_free_forecast.yaml test_aero
 ```
-where $CASE is one of the following:
-- coupled_free_forecast: 2 day tests for atm-ocn-ice coupling 
````
Why were these deleted? These are still actively being used.

JessicaMeixner-NOAA commented:
Have tests been made to make sure we can still run the options that were running before are still running?

rmontuoro commented:
@JessicaMeixner-NOAA - thank you for your comments. This PR includes temporary changes and was issued on @WalterKolczynski-NOAA request to begin integration and testing work needed to include active aerosols in the coupled UFS model. All changes have been tested and do not impact other coupled configurations.

The README was only included to help test this PR and should not be part of the final merge. I'd be happy to remove it or modify it once this review is complete. Thank you for pointing this out!

WalterKolczynski-NOAA commented:
> @JessicaMeixner-NOAA - thank you for your comments. This PR includes temporary changes and was issued on @WalterKolczynski-NOAA request to begin integration and testing work needed to include active aerosols in the coupled UFS model. All changes have been tested and do not impact other coupled configurations.
> 
> The README was only included to help test this PR and should not be part of the final merge. I'd be happy to remove it or modify it once this review is complete. Thank you for pointing this out!

I'd like the README to include instructions for both, with aerosols as the alternate. The scripts already do this with the -a option, it is just covering both in the README.

WalterKolczynski-NOAA commented:
I'm unable to run due to restricted permissions on /scratch1/NCEPDEV/nems/Raffaele.Montuoro/data/NASA/rc

rmontuoro commented:
@WalterKolczynski-NOAA - apologies. I've just fixed the permissions. Please try again.

WalterKolczynski-NOAA committed:
Merge branch 'feature/coupled-crow' into feature/aerosols

WalterKolczynski-NOAA commented:
Resolved conflict due to updated version of ufs_weather_model for coupled model.

WalterKolczynski-NOAA commented:
aerosol_free fails on Hera with a message about the executable size changing during execution:

> *** Error in /scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/ptmp/aerosol_free/2013040100/gfs/fcst.116746/ufs_model': corrupted size vs. prev_size: 0x000000000f9a5dc0 ***

aerosol_firex fails on Hera due to hitting the job time limit

Both seem to actually be completing the forecast (there are log files at FHMAX_GFS for both), so maybe it is hanging at the end for some reason.

[Full log files](https://gist.github.com/WalterKolczynski-NOAA/32aba6543eec7bf721427a9a1558ea76)
Working directories ($DATA):
/scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/ptmp/aerosol_free/2013040100/gfs/fcst.116746
/scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/ptmp/aerosol_firex/2019070100/gfs/fcst.99946

rmontuoro commented:
@WalterKolczynski-NOAA - It looks like an MPI-related issue upon finalize. It also seems both runs completed successfully before the failure occurred. Have you observed this behavior consistently? Could you please grant read permissions to the core dump files generated by the aerosol_free case?

WalterKolczynski-NOAA commented:
I tried multiple times with each.

You should have permission now.

rmontuoro commented:
@WalterKolczynski-NOAA - The failure may be caused by a bug in the Intel OpenMP library. See stack trace below:
```
#0  0x00002b2c6c1153d7 in raise () from /lib64/libc.so.6
#1  0x00002b2c6c116ac8 in abort () from /lib64/libc.so.6
#2  0x000000000384091b in for__signal_handler ()
#3  <signal handler called>
#4  0x00002b2c6c1153d7 in raise () from /lib64/libc.so.6
#5  0x00002b2c6c116ac8 in abort () from /lib64/libc.so.6
#6  0x00002b2c6c157f67 in __libc_message () from /lib64/libc.so.6
#7  0x00002b2c6c15fc37 in malloc_consolidate () from /lib64/libc.so.6
#8  0x00002b2c6c16020e in _int_free () from /lib64/libc.so.6
#9  0x00002b2c5d100908 in _dl_deallocate_tls () from /lib64/ld-linux-x86-64.so.2
#10 0x00002b2c6bcb3ee7 in __free_stacks () from /lib64/libpthread.so.0
#11 0x00002b2c6bcb3fff in __deallocate_stack () from /lib64/libpthread.so.0
#12 0x00002b2c6bcb6053 in pthread_join () from /lib64/libpthread.so.0
#13 0x00002b2c6a0c5dfe in __kmp_reap_worker (th=0x3815e) at ../../src/z_Linux_util.cpp:1122
#14 0x00002b2c6a08e498 in __kmp_internal_end_library (gtid_req=229726) at ../../src/kmp_runtime.cpp:6117
#15 0x00002b2c5d0fe0a8 in _dl_fini () from /lib64/ld-linux-x86-64.so.2
#16 0x00002b2c6c118d39 in __run_exit_handlers () from /lib64/libc.so.6
#17 0x00002b2c6c118d87 in exit () from /lib64/libc.so.6
#18 0x00002b2c6c10155c in __libc_start_main () from /lib64/libc.so.6
#19 0x00000000005c1029 in _start ()
```

The failure no longer occurred after setting
```
export KMP_INIT_AT_FORK=FALSE
```

However, the model still hangs upon finalizing ESMF with `OMP_NUM_THREADS=2`.

WalterKolczynski-NOAA committed:
Make aerosol cases single-threaded for now

There are issues running the aerosol model with OpenMP, so the
number of threads is reduced to one. Attempts to increase the layout
to 12×8 resulted in model crash at initialization for unknown reasons.

All testing only on Hera to this point.

Refs: #305

WalterKolczynski-NOAA committed:
Merge branch 'feature/aerosols' of github.com:rmontuoro/global-workflow into feature/aerosols

WalterKolczynski-NOAA commented:
I've been able to run the aerosol_free case successfully, but the aerosol_firex is still failing during finalize with the following error:

`*** Error in '/scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/ptmp/global-workflow/aerosol_firex/2019070100/gfs/fcst.262680/ufs_model': double free or corruption (fasttop): 0x000000000f8d7a00 ***`

Full logfile: https://gist.github.com/WalterKolczynski-NOAA/75c7c7f38ad51de716f2e1a1683c977f

$DATA: /scratch2/NCEPDEV/ensemble/noscrub/Walter.Kolczynski/ptmp/global-workflow/aerosol_firex/2019070100/gfs/fcst.262680

All the post jobs run fine, since all the output is produced correctly.

WalterKolczynski-NOAA commented:
Actually, ignore the last bit. I think I might've used -c instead of -a, so it wasn't the aerosol versions of stuff.

WalterKolczynski-NOAA commented:
I think I had the wrong model version checked out in the last test and now the firex forecast has completed and post is running. So I may finally have it.

WalterKolczynski-NOAA commented:
Confirmed it works for both aerosol cases now. Currently doing regression test to update so the normal coupled cases work. Already had to add some settings to the chem defaults (commit coming).

rmontuoro commented:
@WalterKolczynski-NOAA - Good to hear. Looking forward to your commits.

WalterKolczynski-NOAA committed:
Add missing defaults for chem settings

Due to the way CROW is currently set up, all of the chem_settings
need to be defined even if they aren't used or relevant. Defaults
for the missing settings have been added by using the string
"not_set" when cplgocart is not true.

This arrangment results in a slightly different error message if
a case with gocart on fails to set one of the necessary settings.
Instead of getting a missing variable exception:

crow.config.exceptions.VariableMissing:
  doc.chem_settings[0]: missing:
    CFGDIR, INPDIR in: CDUMP, Template, MOD, ntdiag

You get a type exception that the type is null with the misleading
suggestion that it should be in a !calc:

doc.chem_settings[0].CFGDIR=null:
  not valid for type 'string'.  Should this have been a !calc?

Refs: #304

WalterKolczynski-NOAA committed:
Merge branch 'feature/coupled-crow' of github.com:NOAA-EMC/global-workflow into feature/aerosols

WalterKolczynski-NOAA committed:
Remove dead setting from fcst config

The KMP_INIT_AT_FORK setting had been tested at one point, but was
never active in committed code because it is not necessary, so the
commented line should've never been committed.

Refs: #304

WalterKolczynski-NOAA committed:
Update README with instructions for coupled and aerosol

Added back in the instructions for the coupled model that were replaced
by the aerosol instructions and combined the two into a single set of
instructions. Both methods need to viable until the two applications are
fully combined.

Refs: #304

WalterKolczynski-NOAA commented:
@JessicaMeixner-NOAA I updated the README to add the coupled instructions back in and make a single reference with instructions for both. Please review if this addresses all your concerns.

WalterKolczynski-NOAA reviewed (approved):
Will hold off on merging until @JessicaMeixner-NOAA approves as well.

WalterKolczynski-NOAA commented:
Also running all three tests again with final version.

JessicaMeixner-NOAA commented:
Thanks for the updates to the read me!

JessicaMeixner-NOAA reviewed (approved):
One comment on the first line of the readme, but all the other changes were made -- thank you! 

Since Walter's test for the coupled side of things work and the readme was updated to include information for both I no longer have any objections.
